codetransformer/assembler.py

```diff
@@ -0,0 +1,99 @@
+import sys
+import types
+from toolz import mapcat
+
+from . import instructions as instrs
+from .code import Code
+
+
+Label = instrs.Label
+
+
+def assemble_function(signature, objs, code_kwargs=None, function_kwargs=None):
+    """TODO
+    """
+    if code_kwargs is None:
+        code_kwargs = {}
+    if function_kwargs is None:
+        function_kwargs = {}
+
+    code_kwargs.setdefault('argnames', list(gen_argnames_for_code(signature)))
+
+    # Default to using the globals of the calling stack frame.
+    function_kwargs.setdefault('globals', sys._getframe(1).f_globals)
+
+    function_kwargs.setdefault('argdefs', tuple(extract_defaults(signature)))
+
+    code = assemble_code(objs, **code_kwargs).to_pycode()
+
+    return types.FunctionType(code, **function_kwargs)
+
+
+def assemble_code(objs, **code_kwargs):
+    """TODO
+    """
+    instrs = resolve_labels(assemble_instructions(objs))
+    return Code(instrs, **code_kwargs)
+
+
+def assemble_instructions(objs):
+    """Assemble a sequence of Instructions or iterables of instructions.
+    """
+    return list(mapcat(_validate_instructions, objs))
+
+
+def resolve_labels(objs):
```

Comment:
In an earlier draft of this I was allowing people to pass `Macro`s as arguments to jumps, and I replaced jumps to macros with jumps to the first instruction here. Now that we have labels, I can't think of a use-case where passing a macro is a significant improvement over passing a label that's right before a label, though it does mean you have to make sure nothing gets put between your label and your macro. One thing we could consider is adding `start` and `end` labels to all macros and have `__iter__` bookend the results of `assemble` with those.

Reply:
I think that adding the `start`/`end` labels is easier to support elsewhere, but we can implement that later.